[App] Support app checkpointing

manskx opened the pull request:
## What does this PR do?


User story:
As a user, I want to save a checkpoint (app state) of my app to a saved file.
As a user, I want to start my app from a saved checkpoint locally on drive 

How to save an app checkpoint?
inside app.py use this hook `should_save_checkpoint()` that controls whether to save the checkpoint or not.

example:

app.py
```python
class SavedApp(L.LightningFlow):

    def __init__(self):
        super().__init__()
        self.text = ""
        self.save_checkpoint = False

    def run(self):
        self.text = "Hello World!"
        self.save_checkpoint = True

    def should_save_checkpoint(self):
        if self.save_checkpoint:
            self.save_checkpoint = False
            return True
        return False
```

checkpoints will be saved in the shared storage in this directory `checkpoints` and will be with the file name `lightningapp_checkpoint_<timestamp>.json`

How to start an app from a saved checkpoint?

Using CLI, starting an app from a checkpoint can be done as follows

```
#localfile  
lightning run app app.py --cloud --checkpoint lightningapp_checkpoint_1665501626.json   
```
 
```
#latest saved checkpoint  
lightning run app app.py --cloud --checkpoint latest   
```

```
# checkpoint name on the drive
lightning run app app.py --cloud --checkpoint lightningapp_checkpoint_1665501626   
```

works locally and on the cloud.

Enable this env var to use the feature `ENABLE_APP_CHECKPOINT`

### Does your PR introduce any breaking changes? If yes, please list them.
None


Docs: https://github.com/Lightning-AI/lightning/pull/15216

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

<!-- In the CHANGELOG, separate each item in the unreleased section by a blank line to reduce collisions -->

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃


cc @borda @awaelchli @ananthsub @ninginthecloud @rohitgr7 @otaj

tchaton commented on src/lightning_app/core/app.py:
```diff
@@ -521,6 +527,83 @@ def state_dict(self) -> t.Dict:
+
+        if not state:
+            raise FileNotFoundError(f"Could not find checkpoint: {checkpoint}")
+        self.load_state_dict(state)
```
This won't work with dynamic components. We need to finish the logic with load_state and get_state.

manskx replied:
Yes, supporting dynamic components might be tricky. Let me try to fix it, otherwise I will add it in another PR.

manskx replied:
One challenge with dynamic works is how to load the component and initialize it from the saved checkpoint, do you have ideas to do it, maybe pickle ? @tchaton

manskx replied:
Or it should be expected from the user to implement `load_state_dict()` and Re-instantiate the dynamic components or have it in a separate hook.

manskx replied:
updated

tchaton reviewed (commented):
Overall, this is a great start, but I would like to see this feature tried out in a real app. I am mitigated on its ability to ensure app can be restored from the checkpoint.
Did you try with a real use case ?

manskx commented:
> Overall, this is a great start, but I would like to see this feature tried out in a real app. I am mitigated on its ability to ensure app can be restored from the checkpoint. Did you try with a real use case ?

I added a testcase with nested dynamic components apps and tested it with a few of the examples like the boring_app and it works fine.

tchaton commented on src/lightning_app/runners/multiprocess.py:
```diff
@@ -4,6 +4,7 @@
 from typing import Any, Callable, Optional, Union
 
 from lightning_app.api.http_methods import _add_tags_to_api, _validate_api
+from lightning_app.core import constants
```
Let's use `ENABLE_APP_CHECKPOINT` directly.

tchaton commented on src/lightning_app/utilities/app_helpers.py:
```diff
@@ -419,7 +419,7 @@ def _set_level(self):
 def _state_dict(flow: "LightningFlow"):
     state = {}
-    flows = [flow] + list(flow.flows.values())
+    flows = [flow] + flow.get_all_children()
```
Note for another PR. Let's name `get_all_children` to `flows`.

manskx replied:
+1 

I'm going to revert this, it's not needed here.

tchaton commented on tests/tests_app/core/test_lightning_app_checkpoint.py:
```diff
@@ -0,0 +1,146 @@
+from lightning_app import LightningApp, LightningFlow, LightningWork  # F401
+
+
+class SavedAppExample(LightningFlow):
```
Could you test out with an app where the components take input arguments? This is usually the main issue.

manskx replied:
added 👍

tchaton commented on tests/tests_app/core/test_lightning_app_checkpoint.py:
```diff
@@ -0,0 +1,146 @@
+        self.work_level_3.run()
+
+    def load_state_dict(self, flow_state, children_states, strict) -> None:
+        self.work_level_3 = DynamicWorkLevel3()
```
Could you try to see if overriding `state_dict` and add some extras required to re-instantiation works well ?

tchaton commented on src/lightning_app/core/constants.py:
```diff
@@ -65,3 +65,5 @@ def get_lightning_cloud_url() -> str:
 
 DEBUG: bool = lightning_cloud.env.DEBUG
 DEBUG_ENABLED = bool(int(os.getenv("LIGHTNING_DEBUG", "0")))
+
```
Let's re-organise the experimental features at the bottom. 

Maybe, we could have a section in the doc about experimental features.

tchaton commented on tests/tests_app/core/test_lightning_app_checkpoint.py:
```diff
@@ -0,0 +1,146 @@
+        super().load_state_dict(flow_state, children_states, strict=strict)
+
+
+class DynamicFlowLevel1(LightningFlow):
```
Could you validate this work with structures like Dict and List ?

tchaton reviewed (commented):
Great progress, starting to look good !

tchaton commented:
Hey @manskx @ethanwharris . Do you believe you could try this feature Flashy. This here the struggle was. Worth to see if he help in real example.

manskx commented:
> Hey @manskx @ethanwharris . Do you believe you could try this feature with Flashy. This is where the struggle was. Worth to see if he help in real example.

will test it with Flashy

awaelchli commented on src/lightning_app/core/app.py:
```diff
@@ -545,6 +629,7 @@ def load_state_dict_from_checkpoint_dir(
         self.load_state_dict(state)
 
     def _dump_checkpoint(self) -> Optional[str]:
+        # TODO: revisit this logic.
```
I think you meant to say that this method should be removed once the checkpointing is no longer experimental

awaelchli commented on src/lightning_app/core/app.py:
```diff
@@ -568,6 +653,32 @@ def _dump_checkpoint(self) -> Optional[str]:
+        saved in the local directory. If the app is running on the cloud, the checkpoint will be saved in the cloud
+        storage.
+
+        :param checkpoint_name: The name of the checkpoint, if not defined, the checkpoint will be saved with the
+            current timestamp.
+        :return: the path of the checkpoint file.
```
Follow this doc style:

```
Args:
    arg1: Description
    arg2: ...

Return:
    Description
```

awaelchli commented on src/lightning_app/core/app.py:
```diff
@@ -568,6 +653,32 @@ def _dump_checkpoint(self) -> Optional[str]:
+        drive = Drive(
+            "lit://checkpoints", component_name="root", allow_duplicates=True, root_folder=self.checkpoint_dir
+        )
+        with _context(ComponentContext.WORK):
```
Why are we pretending to be in a work context? Let's not do this. Can't we just save it directly to the shared path?

lantiga replied:
That's also my question with these work contexts, can you clarify what is the underlying holdup @manskx ?

manskx replied:
It's because the drive doesn't allow the flow to put/get/list files 
https://github.com/Lightning-AI/lightning/blob/e9a6b834370251252d0e6c9767cac8bd7c81407b/src/lightning_app/storage/drive.py#L90

manskx replied:
@tchaton what is the best practice here?

awaelchli commented on src/lightning_app/core/flow.py:
```diff
@@ -757,17 +757,26 @@ def load_state_dict(self, flow_state, children_states, strict) -> None:
+        for structure, state in flow_state["structures"].items():
+            getattr(self, structure).set_state(state)
+
+    def should_save_checkpoint(self):
```
```suggestion
    def should_save_checkpoint(self) -> bool:
```

awaelchli commented on src/lightning_app/runners/cloud.py:
```diff
@@ -111,6 +112,11 @@ def dispatch(
             # Override the cluster ID if provided by the CLI
             app_config.cluster_id = cluster_id
 
+        if ENABLE_APP_CHECKPOINT and self.checkpoint:
+            app_config.checkpoint = self.checkpoint
+        else:
+            app_config.checkpoint = None
```
```suggestion
        if ENABLE_APP_CHECKPOINT and self.checkpoint:
            app_config.checkpoint = self.checkpoint
```
You already added the default to the config

awaelchli commented on src/lightning_app/runners/cloud.py:
```diff
@@ -111,6 +112,11 @@ def dispatch(
             app_config.cluster_id = cluster_id
 
+        if ENABLE_APP_CHECKPOINT and self.checkpoint:
+            app_config.checkpoint = self.checkpoint
```
What is the motivation behind saving this to the app config? Are we accessing it anywhere?

manskx replied:
Checkpoint is a saved state to run the app anywhere from that state without updating the sourcecode. so
when running the app on the cloud, the checkpoint info needs to be sent over. Ideally, you could start the app from the UI with a different name/checkpoint which is considered an app config. 

does it make sense ?

awaelchli commented on src/lightning_app/runners/runtime.py:
```diff
@@ -92,6 +95,7 @@ class Runtime:
     backend: Optional[Union[str, Backend]] = "multiprocessing"
     env_vars: Dict[str, str] = field(default_factory=dict)
     secrets: Dict[str, str] = field(default_factory=dict)
+    checkpoint: str = None
```
```suggestion
    checkpoint: Optional[str] = None
```

awaelchli commented on src/lightning_app/utilities/app_helpers.py:
```diff
@@ -430,6 +430,7 @@ def _state_dict(flow: "LightningFlow"):
     return state
 
 
+# TODO: This helper doesn't seem to be used anywhere
```
@tchaton keep or remove? or fix?

awaelchli commented on src/lightning_app/utilities/packaging/app_config.py:
```diff
@@ -15,10 +15,12 @@ class AppConfig:
 
     name: str = field(default_factory=get_unique_name)
     cluster_id: Optional[str] = field(default=None)
+    checkpoint: Optional[str] = None  # TODO: might be moved to appspec.
```
What are the considerations here? Is it really needed to mention this in the source code?

manskx replied:
let's remove it

awaelchli commented on tests/tests_app/core/test_lightning_flow.py:
```diff
@@ -674,109 +669,3 @@ def run(self):
     Flow().run()
-
-
-class WorkReload(LightningWork):
```
Could you explain which of these tests have moved and where, and which have been removed and why?

manskx replied:
I removed these tests because they test these two methods `_state_dict` and `_load_state_dict` which are not used and have some bugs, that's why I put a comment above to remove them.

awaelchli commented on src/lightning_app/core/app.py:
```diff
@@ -520,6 +526,84 @@ def state_dict(self) -> Dict:
+    @classmethod
+    def _load_checkpoint_from_json_file(cls, path: str) -> None:
+        checkpoint_dict = json.load(open(path))
+        # TODO: Checkpoint versioning and compatibility checks
```
Where will this TODO be addressed?
- In this PR?
- Not in this PR? -> Let's open a GitHub issue

manskx replied:
created a GH issue for it: https://github.com/Lightning-AI/lightning/issues/15448

awaelchli commented on src/lightning_app/core/app.py:
```diff
@@ -390,6 +392,7 @@ def run_once(self):
         self._update_layout()
         self.maybe_apply_changes()
 
+        # TODO: This should be removed
```
Let's have the TODO mention when to remove it. I think you meant to say that it should be removed once the new logic is no longer experimental.

awaelchli commented on src/lightning_app/core/app.py:
```diff
@@ -520,6 +526,84 @@ def state_dict(self) -> Dict:
+
+    def _get_checkpoint_if_available_locally(self, checkpoint: str) -> Optional[dict]:
+        checkpoint_dict = None
+        if os.path.exists(checkpoint):
+            checkpoint_dict = self._load_checkpoint_from_json_file(checkpoint)
+        if os.path.exists(os.path.join(self.checkpoint_dir, checkpoint)):
```
I don't understand why there are two possible ways to load here. Isn't checkpoint a path?

manskx replied:
I want to allow either using the full path or only the name of the checkpoint.

awaelchli commented on src/lightning_app/core/app.py:
```diff
@@ -520,6 +526,84 @@ def state_dict(self) -> Dict:
+
+        return checkpoint_dict
+
+    def _get_checkpoint_if_available_on_drive(self, checkpoint: str) -> Optional[dict]:
```
Can this and perhaps the one above become utility functions? We could construct the path here in the app and then call the utility function with the already constructed path.

manskx replied:
Why moving it to the utility function? I am not sure if it should be reused outside of the app class.

awaelchli commented on src/lightning_app/core/app.py:
```diff
@@ -520,6 +526,84 @@ def state_dict(self) -> Dict:
+
+    def _get_checkpoint_if_available_on_drive(self, checkpoint: str) -> Optional[dict]:
+        os.makedirs(self.checkpoint_dir, exist_ok=True)
+        drive = Drive("lit://checkpoints", component_name="root", root_folder=self.checkpoint_dir)
+
+        with _context(ComponentContext.WORK):
+            found_checkpoints = drive.list()
```
Same comment here about the use of _context(work). We didn't design it to be used this way. @tchaton what is your recommendation?

awaelchli commented on src/lightning_app/core/app.py:
```diff
@@ -520,6 +526,84 @@ def state_dict(self) -> Dict:
+                os.path.join(self.checkpoint_dir, checkpoint_drive_path)
+            )
+        else:
+            raise FileNotFoundError(
+                f"Could not read checkpoint file: {checkpoint_drive_path} although it was found on the drive."
+            )
```
How exactly can we say that we "found" it? The conditions in the if-statements above do not hold, otherwise we wouldn't end up in the else. 

Can we mention an action that the user can take?

manskx replied:
👍  removed this else completely, it doesn't make sense

awaelchli commented on src/lightning_app/core/app.py:
```diff
@@ -520,6 +526,84 @@ def state_dict(self) -> Dict:
+
+        checkpoint_drive_path = ""
+
+        if checkpoint == "latest":
+            checkpoint_filenames_regex = r"lightningapp_checkpoint_([0-9]{10}).json"
+            r = re.compile(checkpoint_filenames_regex)
+            filtered_list = list(filter(r.search, found_checkpoints))
+            checkpoint_drive_path = max(filtered_list, key=lambda i: int(r.search(i).group(1))) if filtered_list else ""
+        else:
+            for cp in found_checkpoints:
+                if checkpoint in cp:
+                    checkpoint_drive_path = cp
+                    break
```
Can we make this a function? This would make it much easier to unit-test the cases.

manskx replied:
added in an utility function

awaelchli commented:
I recommend adding docs directly to feature PRs. It helps the reviewer understand what is being implemented. Plus it is easier to check whether the implementation matches what is explained in the docs and the other way around. This is a personal preference but it has been very useful in the past.

lantiga commented on src/lightning_app/core/constants.py:
```diff
@@ -52,19 +52,24 @@ def get_lightning_cloud_url() -> str:
 
 # EXPERIMENTAL: ENV VARIABLES TO ENABLE MULTIPLE WORKS IN THE SAME MACHINE
 DEFAULT_NUMBER_OF_EXPOSED_PORTS = int(os.getenv("DEFAULT_NUMBER_OF_EXPOSED_PORTS", "50"))
```
these ENV vars are quite a few at this point, I suggest we prepend them with a prefix (`LIGHTNING`, `LAI` or such).

manskx replied:
I updated the env var to start with LAI prefix. there's an issue here to update all the other envs: https://github.com/Lightning-AI/lightning/issues/15388

lantiga commented on tests/tests_app/core/test_lightning_app_checkpoint.py:
```diff
@@ -0,0 +1,256 @@
+        assert text_from_saved_state == "Hello Work Level 3!"
+
+        # now we want to update this state with the new value of text and re-instantiate the work with the required args
+        children_states["work_level_3"]["vars"]["text"] = "Hello Work Level 3 Updated!"
```
Are we expecting developers implementing `load_state_dict` to manually dig into states this way?

I think we should provide a lightweight API that allows to get var values from the state, otherwise we are coupling `load_state_dict` implementations with the current structure of state. It could be as lightweight as just wrapping the dict in a wrapper class that provides `.` access to "vars" from the state.

I'm good with keeping it this way right now, but I'd open an issue right after merging because we'll need to do something about it in the near future.

manskx replied:
yes, this will be nice. 
created an issue here: https://github.com/Lightning-AI/lightning/issues/15475

lantiga reviewed (commented):
Good, let's get this past the finish line

manskx commented:
> 

@awaelchli I have closed the docs PR and added the docs to this one.

manskx commented on src/lightning_app/core/constants.py:
```diff
@@ -52,19 +52,24 @@ def get_lightning_cloud_url() -> str:
+ENABLE_MULTIPLE_WORKS_IN_DEFAULT_CONTAINER = bool(
+    int(os.getenv("ENABLE_MULTIPLE_WORKS_IN_DEFAULT_CONTAINER", "0"))
+)  # Note: This is disabled for the time being.
+ENABLE_APP_CHECKPOINT = bool(int(os.getenv("ENABLE_APP_CHECKPOINT", "0")))
```
updated

manskx commented:
@lantiga @awaelchli @tchaton the PR is updated with the addressed comments, I also added the docs in the same PR. 

Could you please have another look ?

awaelchli commented on docs/source-app/glossary/app_checkpointing.rst:
```diff
@@ -0,0 +1,113 @@
+            # do stuff
+            pass
+
+        def should_save_checkpoint(self):
+            return True
```
Can any node of the flow tree define it or should we restrict it to the root only?

manskx replied:
Good question. It's only the root. I will add a comment here, and maybe enforce it with an error message in another PR.

awaelchli commented on docs/source-app/glossary/app_checkpointing.rst:
```diff
@@ -0,0 +1,113 @@
+How to save a checkpoint
+************************
+
+Saving a checkpoint is simple. In main flow if your lightning app, implement `should_save_checkpoint()` that returns True when you want to save a checkpoint.
```
```suggestion
Saving a checkpoint is simple. In the main flow of your Lightning App, implement ``should_save_checkpoint()`` that returns ``True`` when you want to save a checkpoint.
```

awaelchli commented on docs/source-app/glossary/app_checkpointing.rst:
```diff
@@ -0,0 +1,113 @@
+        def should_save_checkpoint(self):
+            return True
+
+When `should_save_checkpoint()` returns True, Lightning will save the state of the app to a checkpoint file. The checkpoint file will be saved in lightning shared storage at `/checkpoints` and will be named `lightningapp_checkpoint_<timestamp>.json`.
```
```suggestion
When ``should_save_checkpoint()`` returns ``True``, Lightning will save the state of the app to a checkpoint file. The checkpoint file will be saved in Lightning shared storage at ``/checkpoints`` and will be named ``lightningapp_checkpoint_<timestamp>.json``.
```

awaelchli replied:
Also, /checkpoints (root) is not correct, right? It probably means ./checkpoints (cwd)?

manskx replied:
Yes, I would say ``lit://checkpoints`

"The checkpoint file will be saved in Lightning [Drive Storage](https://lightning.ai/docs/stable/glossary/storage/drive.html) at `lit://checkpoints` "

awaelchli commented on docs/source-app/glossary/app_checkpointing.rst:
```diff
@@ -0,0 +1,113 @@
+
+
+*************************************
+How to an app start from a checkpoint
```
There seems to be a problem with this title

awaelchli commented on docs/source-app/glossary/app_checkpointing.rst:
```diff
@@ -0,0 +1,113 @@
+    lightning run app app.py --checkpoint lightningapp_checkpoint_1665501626.json
+
+
+2. The name of the checkpoint file in lightning shared storage
```
```suggestion
2. The name of the checkpoint file in Lightning shared storage.
```

What is shared storage, does user know that?

awaelchli commented on docs/source-app/glossary/app_checkpointing.rst:
```diff
@@ -0,0 +1,113 @@
+    lightning run app app.py --checkpoint lightningapp_checkpoint_1665501688
+
+
+3. keyword `latest` to start from the latest checkpoint file in lightning shared storage.
```
```suggestion
3. The keyword `latest` to start from the latest checkpoint file in Lightning shared storage.
```

awaelchli commented on docs/source-app/glossary/app_checkpointing.rst:
```diff
@@ -0,0 +1,113 @@
+
+When starting an app from a saved checkpoint, it has to be compatible with the app code.
+This means that the app code has to be able to load the saved state and update all app components with the saved state.
+If the checkpoint contains a state of a component that is not yet instantiated in the component init(), the app won't start and will raise an exception.
```
```suggestion
If the checkpoint contains a state of a component that is not yet instantiated in the component ``__init__()``, the app won't start and will raise an exception.
```

awaelchli commented on docs/source-app/glossary/app_checkpointing.rst:
```diff
@@ -0,0 +1,113 @@
+This means that the app code has to be able to load the saved state and update all app components with the saved state.
+If the checkpoint contains a state of a component that is not yet instantiated in the component init(), the app won't start and will raise an exception.
+
+To control how the checkpoint is loaded and create the missing components, implement `load_state_dict()` in your app. You can also implement any migration logic needed to read or update the checkpoint before loading it or instantiate component attributes that are not yet instantiated in the component init().
```
```suggestion
To control how the checkpoint is loaded and create the missing components, implement ``load_state_dict()`` in your app. You can also implement any migration logic needed to read or update the checkpoint before loading it or instantiate component attributes that are not yet instantiated in the component ``__init__()``.
```

awaelchli commented on docs/source-app/glossary/app_checkpointing.rst:
```diff
@@ -0,0 +1,113 @@
+            )
+
+
+.. note:: If you see this exception "The component <component_name> wasn't instantiated for the component root", it means that the checkpoint is not compatible with the app code and you need to implement `load_state_dict()` and make sure that all components in the checkpoint are instantiated.
```
Do you see any way we could actually give this specific advice directly in such an error message (possibly new PR)

manskx replied:
I like it. I would add it to this PR.

awaelchli reviewed (approved):
reviewed docs

Borda commented:
> @tchaton or @awaelchli can you pls finish it? rabbit

@lantiga how do you think? :chipmunk:

lantiga commented:
This is a valuable contribution, but given what's in flux right now I propose to keep on-hold (we can keep as is or close for now and reopen in the future)
